cou/apps/auxiliary_subordinate.py

```diff
@@ -33,6 +44,59 @@ def _check_ovn_support(self) -> None:
         """
         OVNSubordinate._validate_ovn_support(self.workload_version)
 
+    def _check_auto_restarts(self) -> None:
+        """No-op, skip check auto restarts option.
+
+        This method override the parent class's `_check_auto_restarts()` method
+        because the parent class's will raise an `ApplicationError` if
+        `enable-auto-restarts` is `True`.
+        """
+
+    def pre_upgrade_steps(
+        self, target: OpenStackRelease, units: Optional[list[Unit]]
+    ) -> list[PreUpgradeStep]:
+        """Pre Upgrade steps planning.
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :param units: Units to generate upgrade plan
+        :type units: Optional[list[Unit]]
+        :return: List of pre upgrade steps.
+        :rtype: list[PreUpgradeStep]
+        """
+        # OVN subordinate charm does not have upgrade_step; the "upgrade" is
+        # done during the pre-upgrade step, so we need to put the run deferred
+        # hook step before pre-upgrade step.
+        steps = []
+        if self.config.get("enable-auto-restarts", {}).get("value") is False:
+            steps.extend(
+                self.get_run_deferred_hooks_and_restart_pre_upgrade_step(self.subordinate_units)
+            )
```

Comment:
You need to receive specific units from a higher level. Upgrading all units at once here you'll create a possible data-plane network outage if all nova-compute units restarts the ovn services at the same time

Reply:
@chanchiwai-ray the deferred restarts mechanism was introduced to avoid restarting all units at once and allow operators to do a rolling restart. This is to avoid generating downtime when quorum in the cluster is lost.  When an operator runs manual upgrades, they will run deferred hooks / service restarts on one unit, wait until the cluster is back to full redundancy, and then move on to the next one.
Are we running deferred hooks / service restarts all at once in parallel for all services we are supporting? If so we need to revisit those codepaths.
Also note that for ovn-chassis things are a _little_ more relaxed: doing one at a time is fine, but I think it would also be fine to restart all ovn-chassis belonging to empty hypervisors (because the blast radius of those restarts is all the VMs running on that specific host). If we're dealing with ovn-chassis as a subordinate for anything other than nova-compute though, I suspect a strict one at a time strategy would be the only safe one.

Reply:
ok thinking about this some more...
In order not to go crazy with overcomplicating things about rolling upgrades, we should think in terms of classes of applications:
* pure control plane apps: downtime doesn't matter → restart whatever you want without a care for quorum
* empty hypervisors or hypervisors that the user told us to go ahead and upgrade anyway: downtime doesn't matter → quorum isn't a thing here, and you can restart whatever and whenever
* rabbitmq: it's a control plane app, sort of. If it stays down for too long neutron won't be happy and data plane will be affected → handle deferred restarts carefully as to never lose quorum
* ovn-chassis as a subordinate of octavia → if restarting it only causes downtime in the control plane (i.e. pushing updates to amphorae) then there is no need for rolling restarts, if on the other hand it causes LBs to lose packets we should strive to maintain quorum

Reply:
> Are we running deferred hooks / service restarts all at once in parallel for all services we are supporting? If so we need to revisit those codepaths.

No, the substeps are not run in parallel, it's run unit-by-unit until the action completes.

- Refresh the ovn-chassis without restarting the service
  - restart ovn-chassis unit 1 until it's done
  - restart ovn-chassis unit 2 until it's done
  - ...